src/interpreter.ts

```diff
@@ -480,16 +487,37 @@ export const evaluators: { [nodeType: string]: Evaluator<es.Node> } = {
     return yield* evaluate(node.expression, context)
   },
   *ReturnStatement(node: es.ReturnStatement, context: Context) {
-    if (node.argument) {
-      if (node.argument.type === 'CallExpression') {
-        const callee = yield* evaluate(node.argument.callee, context)
-        const args = yield* getArgs(context, node.argument)
-        return new TailCallReturnValue(callee, args, node.argument)
+    let return_expression = node.argument
+    if (!return_expression) {
+      return new ReturnValue(undefined)
+    }
+
+    // If we have a conditional expression, reduce it until we get something else
+    while (return_expression.type === 'LogicalExpression' ||
+      return_expression.type === 'ConditionalExpression') {
+      if (return_expression.type === 'LogicalExpression') {
+        return_expression = transformLogicalExpression(return_expression)
+      }
+      const test = yield* evaluate(return_expression.test, context)
```

Comment:
this part looks like duplication of code from the IfStatement. Maybe can refactor out into a function just returning the resulting node, and use the refactored function here in ReturnStatement, and evaluate the result of the refactored function inside IfStatement? Something like 

```javascript

function getResultantNodeFromIf(args) {
  return node.consequent or node.alternate;
}

ReturnStatement() {
while() {
 ...
 return_expression = getResultantNodeFromIf(args);
 ...
}
}

IfStatement() {
  yield* evaluate(getResultantNodeFromIf(args));
}

And return_expression should be returnExpression instead.

Reply:
Ah, you are right. Thanks, shall refactor now.

Reply:
Can't deduplicate this code as I run into a typescript bug. Can't find a way to resolve...